app/xkbevd/xkbevd.man
.\" $XdotOrg: $
.\" $Xorg: xkbevd.man,v 1.3 2000/08/17 19:54:49 cpqbld Exp $
.\"
.\" $XFree86: xc/programs/xkbevd/xkbevd.man,v 1.7 2001/01/27 18:21:16 dawes Exp $
.\"
.TH XKBEVD 1 __xorgversion__
.SH NAME
xkbevd \- XKB event daemon
.SH SYNOPSIS
.B xkbevd
[ options ]
.SH DESCRIPTION
.PP
This command is very raw and is therefore only partially implemented;  we 
present it here as a rough prototype for developers, not as a general purpose 
tool for end users.  Something like this might make a suitable replacement
for xev;  I'm not signing up, mind you, but it's an interesting idea.
.PP
The
.I xkbevd
event daemon listens for specified XKB events and executes requested commands
if they occur.   The configuration file consists of a list of event 
specification/action pairs and/or variable definitions.   
.PP
An event specification consists of a short XKB event name followed by a 
string or identifier which serves as a qualifier in parentheses;  empty 
parenthesis indicate no qualification and serve to specify the default 
command which is applied to events which do not match any of the other 
specifications.  The interpretation of the qualifier depends on the type 
of the event:  Bell events match using the name of the bell, message events 
match on the contents of the message string and slow key events accept 
any of \fIpress\fP, \fIrelease\fP, \fIaccept\fP, or \fIreject\fP.   No 
other events are currently recognized.
.PP
An action consists of an optional keyword followed by an optional string 
argument.  Currently, \fIxkbev\fP recognizes the actions: \fInone\fP, 
\fIignore\fP, \fIecho\fP, \fIprintEvent\fP, \fIsound\fP, and \fIshell\fP.
If the action is not specified, the string is taken as the name of a sound
file to be played unless it begins with an exclamation point, in which case
it is taken as a shell command.
.PP
Variable definitions in the argument string are expanded with fields from
the event in question before the argument string is passed to the action
processor.   The general syntax for a variable is 
either $\fIc\P or $(\fIstr\fP), where \fIc\fP is a single character and
\fIstr\fP is a string of arbitrary length.  All parameters have both 
single-character and long names.    
.PP
The list of recognized parameters varies from event to event and is too long 
to list here right now.   This is a developer release anyway, so you can
be expected to look at the source code (evargs.c is of particular interest).
.PP
The \fIignore\fP, \fIecho\fP, \fIprintEvent\fP, \fIsound\fP,and \fIshell\fP
actions do what you would expect commands named \fIignore\fP, \fIecho\fP,
\fIprintEvent\fP, \fIsound\fP, and \fIshell\fP to do, except that the sound
command has only been implemented and tested for SGI machines.   It launches
an external program right now, so it should be pretty easy to adapt, 
especially if you like audio cues that arrive about a half-second after you
expect them.
.PP
The only currently recognized variables are \fIsoundDirectory\fP and
\fIsoundCmd\fP.  I'm sure you can figure out what they do.
.SH OPTIONS
.TP 8
.B \-help
Prints a usage message that is far more up-to-date than anything in this
man page.
.TP 8
.B \-cfg \fIfile\fP
Specifies the configuration file to read.   If no configuration file is
specified, \fIxkbevd\fP looks for ~/.xkb/xkbevd.cf and $(LIBDIR)/xkb/xkbevd.cf
in that order.
.TP 8
.B \-sc\ \fIcmd\fP
Specifies the command used to play sounds.
.TP 8
.B \-sd\ \fIdirectory\fP
Specifies a top-level directory for sound files.
.TP 8
.B \-display\ \fIdisplay\fP
Specifies the display to use.  If not present, \fIxkbevd\fP uses $DISPLAY.
.TP 8
.B \-bg
Tells \fIxkbevd\fP to fork itself (and run in the background).
.TP 8
.B \-synch
Forces synchronization of all X requests.  Slow.
.TP 8
.B \-v
Print more information, including debugging messages.   Multiple 
specifications of \fI-v\fP cause more output, to a point.
.SH "SEE ALSO"
X(__miscmansuffix__)
.SH COPYRIGHT
Copyright 1995, Silicon Graphics Computer Systems
Copyright 1995, 1998  The Open Group
.br
See \fIX(__miscmansuffix__)\fP for a full statement of rights and permissions.
.SH AUTHOR
Erik Fortune, Silicon Graphics


xserver/ChangeLog
commit 439c58849304907900e4dc7429aedb0192749c02
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Oct 11 22:03:32 2009 -0700

    Bump version to 1.6.5
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7f2f71da5125ae01a8279a4886090a711100d671
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Sep 28 14:18:45 2009 +1000

    ephyr: if -parent is given, check for a trailing -screen. (#24144)
    
    If -parent is given, don't open up a new window if -screen is given as well.
    The commandline option -screen allows to set the depth of the embedded
    Xephry instance, even though width and height are autoscaled on -parent.
    
    This patch checks for a -screen parameter after -parent and - if one is
    found - delays initializing the screen. The parent window id is stored
    temporarily but re-set after a -screen argument.
    The following command is thus valid:
    
    Xephyr -parent 1234 -screen 640x480@8 -screen 1024x768
    
    It embeds the first 8-bit screen into window 1234 and opens up a new window
    for the second screen. Multiple parent arguments are possible, the screens
    are embedded in-order.
    
    X.Org Bug 24144 <http://bugs.freedesktop.org/show_bug.cgi?id=24144>
    
    Tested-by: Vic Lee
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    (cherry picked from commit 19be992d9dc542b61fa3f4fd32a09071c9e64880)
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c07b2368a7dedb66455eea313c06859afa2e1a9c
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 1 15:24:58 2009 -0700

    Bump to version 1.6.4.901
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit fc31f76b6ff735e4a4006c5454c4d496ee8b154a
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 30 11:40:19 2009 -0700

    Re-fix DGA removal.
    
    Removing DGA ended up breaking any drivers calling into the old
    xf86DiDGAInit function as it tried to see if DGA was already enabled
    and ended up crashing if the VT wasn't completely initialized. Oops.
    
    Also, if the driver initializes DGA itself, have the DiDGA
    initialization overwrite that information as the DiDGA code will call
    ReInit on mode detect.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    (cherry picked from commit db98b26ee145f70e732e2cf4a6ac3de77fdf4adc)

commit 8cf659f3a5914369a2137ac17b689e5a9fe9ca27
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 27 19:21:45 2009 -0700

    Bump to version 1.6.4
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a26fd1a6d61507b3e69a04d6f6c192a6ec363c5c
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Wed Sep 16 15:46:55 2009 +1000

    Don't send events through the master if the device has SendCoreEvents off.
    
    In server 1.6, all devices are attached to the master device (VCP or VCK).
    Sending an event through the master device means the device is sending core
    events. If a device is configured as SendCoreEvents, just send through the
    device, not through the master.
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

commit 507e57381fea6334f7dc8da6925e53d2c76fddcb
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 18 21:12:17 2009 -0700

    xfree86/modes: Remove all framebuffer support from DGA
    
    This removes all rendering and mapping code from xf86DiDGA, leaving
    just mode setting and raw input device access. The mapping code didn't
    have the offset within /dev/mem for the frame buffer and the pixmap
    support assumed that the framebuffer was never reallocated.
    (cherry picked from 0b7c6c728c2e2d8433a188315cc591308a89cd85)
    
    Signed-off-by: Keith Packard <keithp@keithp.com>
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

commit c3d182a47902d02f9e64f933b8565ae336f73f54
Author: Rémi Cardona <remi@gentoo.org>
Date:   Mon Sep 14 17:09:59 2009 +0200

    dix: append "built-ins" to the font path in SetDefaultFontPath
    
    49b93df8a3002db7196aa3fc1fd8dca1c12a55d6 made the hard dependency on
    a "fixed" font go away but only Xorg could use the built-ins fonts by
    default.
    
    With this commit, all DDXs get "built-ins" appended to their FontPath, not
    just Xorg.
    
    Tested with Xorg, Xvfb and Xnest.
    (cherry picked from commit f56cbe1ef24415d0142b9a7d0ab0a031069ccb52)
    
    Signed-off-by: Rémi Cardona <remi@gentoo.org>
    Signed-off-by: Jon TURNEY <jon.turney@dronecode.org.uk>
    Tested-by: Jon TURNEY <jon.turney@dronecode.org.uk>
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9bc4a69040493e589a3811ca5e085e323438996d
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Wed Sep 9 15:48:33 2009 +1000

    fbdevhw: Test for graphics:fb%d as well as graphics/fb%d
    
    Apparently the kernel can't decide on an API to expose to userspace, so
    let's just try both in the hope that one will work.
    
    Signed-off-by: Daniel Stone <daniel@fooishbar.org>
    Acked-by: Michel Dänzer <michel@daenzer.net>
    (cherry picked from commit f4350c66b493d63fa06be87caa958d7033232ea4)

commit d78669c92d751464bc3c8f79f2aa21ca0951e3a9
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Sep 22 12:09:48 2009 -0700

    render: return the supported version rather than just passing the proto's version
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f5d1da499c3d64f0a4ac217eb19911b8e35cb49b
Author: Michel Dänzer <daenzer@vmware.com>
Date:   Tue Jun 23 16:45:39 2009 +0200

    dri2: Don't crash if pPriv is NULL.
    (cherry picked from commit df597709d71f47b8516e27c6fb1bfffd59de5e48)
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 468787bdd2c34f9edf46da76b6357034c216fa7b
Author: Richard Hughes <richard@hughsie.com>
Date:   Fri Aug 14 11:44:35 2009 +0100


app/xkbevd/xkbevd.c
/* $Xorg: xkbevd.c,v 1.4 2000/08/17 19:54:49 cpqbld Exp $ */
/************************************************************
 Copyright (c) 1995 by Silicon Graphics Computer Systems, Inc.

 Permission to use, copy, modify, and distribute this
 software and its documentation for any purpose and without
 fee is hereby granted, provided that the above copyright
 notice appear in all copies and that both that copyright
 notice and this permission notice appear in supporting
 documentation, and that the name of Silicon Graphics not be 
 used in advertising or publicity pertaining to distribution 
 of the software without specific prior written permission.
 Silicon Graphics makes no representation about the suitability 
 of this software for any purpose. It is provided "as is"
 without any express or implied warranty.
 
 SILICON GRAPHICS DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS 
 SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY 
 AND FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT SHALL SILICON
 GRAPHICS BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL 
 DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, 
 DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE 
 OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION  WITH
 THE USE OR PERFORMANCE OF THIS SOFTWARE.

 ********************************************************/
/* $XFree86: xc/programs/xkbevd/xkbevd.c,v 3.8 2001/01/17 23:46:09 dawes Exp $ */

#define	DEBUG_VAR xkbevdDebug
#include <X11/Xosdefs.h>
#include <stdlib.h>
#include "xkbevd.h"


#define	lowbit(x)	((x) & (-(x)))

/***====================================================================***/

#ifndef DFLT_XKBEVD_CONFIG
#define DFLT_XKBEVD_CONFIG "%s/.xkb/xkbevd.cf"
#endif /* DFLT_XKBEVD_CONFIG */

#ifndef DFLT_XKB_CONFIG_ROOT
#define	DFLT_XKB_CONFIG_ROOT "/usr/X11R6/lib/xkb"
#endif

#ifndef DFLT_SYS_XKBEVD_CONFIG
#define DFLT_SYS_XKBEVD_CONFIG "%s/xkbevd.cf"
#endif /* DFLT_SYS_XKBEVD_CONFIG */

#ifndef DFLT_SOUND_CMD
#define	DFLT_SOUND_CMD "/usr/sbin/sfplay -q"
#endif /* DFLT_SOUND_CMD */

#ifndef DFLT_SOUND_DIR
#define	DFLT_SOUND_DIR "/usr/share/data/sounds/prosonus/"
#endif /* DFLT_SOUND_DIR */

/***====================================================================***/

static char *	dpyName=	NULL;
Display *	dpy=		NULL;
static char *	cfgFileName=	NULL;
int		xkbOpcode=	0;
int		xkbEventCode=	0;
Bool		detectableRepeat= False;

static
CfgEntryPtr	config=		NULL;
static
unsigned long	eventMask=	0;

static Bool	synch=		False;
static int	verbose=	0;
static Bool	background=	False;

static char *	soundCmd=	NULL;
static char *	soundDir=	NULL;

XkbDescPtr	xkb=		NULL;

/***====================================================================***/

#define	M(m)	fprintf(stderr,(m))
#define	M1(m,a)	fprintf(stderr,(m),(a))

static void
Usage(int argc, char *argv[])
{
    M1("Usage: %s [options]...\n",argv[0]);
    M("Legal options:\n");
    M("-?,-help             Print this message\n");
    M("-cfg <file>          Specify a config file\n");
    M("-sc <cmd>            Specify the command to play sounds\n");
    M("-sd <dir>            Specify the root directory for sound files\n");
    M("-d[isplay] <dpy>     Specify the display to watch\n");
    M("-bg                  Run in background\n");
    M("-synch               Force synchronization\n");
    M("-v                   Print verbose messages\n");
    return;
}

/***====================================================================***/

static Bool
parseArgs(int argc, char *argv[])
{
register int i;

    for (i=1;i<argc;i++) {
	if (strcmp(argv[i],"-bg")==0) {
	    background= True;
	}
	else if (strcmp(argv[i],"-cfg")==0) {
	    if (i>=(argc-1)) {
		uError("No configuration file specified on command line\n");
		uAction("Trailing %s argument ignored\n",argv[i]);
	    }
	    else {
		char *name= argv[++i];
		if (cfgFileName!=NULL) {
		    if (uStringEqual(cfgFileName,name)) 
			uWarning("Config file \"%s\" specified twice!\n");
		    else {
			uWarning("Multiple config files on command line\n");
			uAction("Using \"%s\", ignoring \"%s\"\n",name,
								cfgFileName);
		    }
		}
		cfgFileName= name;
	    }
	}
	else if ((strcmp(argv[i],"-d")==0)||(strcmp(argv[i],"-display")==0)) {
	    if (i>=(argc-1)) {
		uError("No display specified on command line\n");
		uAction("Trailing %s argument ignored\n",argv[i]);
	    }
	    else {
		char *name= argv[++i];
		if (dpyName!=NULL) {
		    if (uStringEqual(dpyName,name)) 
			uWarning("Display \"%s\" specified twice!\n");
		    else {
			uWarning("Multiple displays on command line\n");
			uAction("Using \"%s\", ignoring \"%s\"\n",name,
								dpyName);
		    }
		}
		dpyName= name;
	    }
	}
	else if (strcmp(argv[i],"-sc")==0) {
	    if (i>=(argc-1)) {
		uError("No sound command specified on command line\n");
		uAction("Trailing %s argument ignored\n",argv[i]);
	    }
	    else {
		char *name= argv[++i];
		if (soundCmd!=NULL) {
		    if (uStringEqual(soundCmd,name)) 
			uWarning("Sound command \"%s\" specified twice!\n");
		    else {
			uWarning("Multiple sound commands on command line\n");
			uAction("Using \"%s\", ignoring \"%s\"\n",name,
								soundCmd);
		    }
		}
		soundCmd= name;
	    }
	}
	else if (strcmp(argv[i],"-sd")==0) {
	    if (i>=(argc-1)) {
		uError("No sound directory specified on command line\n");
		uAction("Trailing %s argument ignored\n",argv[i]);
	    }
	    else {
		char *name= argv[++i];
		if (soundDir!=NULL) {
		    if (uStringEqual(soundDir,name)) 
			uWarning("Sound directory \"%s\" specified twice!\n");
		    else {
			uWarning("Multiple sound dirs on command line\n");
			uAction("Using \"%s\", ignoring \"%s\"\n",name,
								soundDir);
		    }
		}
		soundDir= name;
	    }
	}
	else if ((strcmp(argv[i],"-synch")==0)||(strcmp(argv[i],"-s")==0)) {
	    synch= True;
	}
	else if (strcmp(argv[i],"-v")==0) {
	    verbose++;
	}
	else if ((strcmp(argv[i],"-?")==0)||(strcmp(argv[i],"-help")==0)) {
	    Usage(argc,argv);
	    exit(0);
	}
	else {
	    uError("Unknown flag \"%s\" on command line\n",argv[i]);
	    Usage(argc,argv);
	    return False;
	}
    }
    if (background == False) {
	eventMask = XkbAllEventsMask;
	verbose++;
    }

    return True;
}

static Display *
GetDisplay(char *program, char *dpyName, int *opcodeRtrn, int *evBaseRtrn)
{
int	mjr,mnr,error;
Display	*dpy;

    mjr= XkbMajorVersion;
    mnr= XkbMinorVersion;
    dpy= XkbOpenDisplay(dpyName,evBaseRtrn,NULL,&mjr,&mnr,&error);
    if (dpy==NULL) {
	switch (error) {
	    case XkbOD_BadLibraryVersion:
		uInformation("%s was compiled with XKB version %d.%02d\n",
				program,XkbMajorVersion,XkbMinorVersion);
		uError("X library supports incompatible version %d.%02d\n",
				mjr,mnr);
		break;
	    case XkbOD_ConnectionRefused:
		uError("Cannot open display \"%s\"\n",dpyName);
		break;
	    case XkbOD_NonXkbServer:
		uError("XKB extension not present on %s\n",dpyName);
		break;
	    case XkbOD_BadServerVersion:
		uInformation("%s was compiled with XKB version %d.%02d\n",
				program,XkbMajorVersion,XkbMinorVersion);
		uError("Server %s uses incompatible version %d.%02d\n",
				dpyName,mjr,mnr);
		break;
	    default:
		uInternalError("Unknown error %d from XkbOpenDisplay\n",error);
	}
    }
    else if (synch)
	XSynchronize(dpy,True);
    if (opcodeRtrn)
	XkbQueryExtension(dpy,opcodeRtrn,evBaseRtrn,NULL,&mjr,&mnr);
    return dpy;
}

/***====================================================================***/

void
InterpretConfigs(CfgEntryPtr cfg)
{
char *		name;
unsigned	priv= 0;

    config= cfg;
    while (cfg!=NULL) {
	name= cfg->name.str;
	if (cfg->entry_type==VariableDef) {
	    if (uStrCaseEqual(name,"sounddirectory")||
					uStrCaseEqual(name,"sounddir")) {
		if (soundDir==NULL) {
		    soundDir= cfg->action.text;
		    cfg->name.str= NULL;
		    cfg->action.text= NULL;
		}
	    }
	    else if (uStrCaseEqual(name,"soundcommand")||
				uStrCaseEqual(name,"soundcmd")) {
		if (soundCmd==NULL) {
		    soundCmd= cfg->action.text;
		    cfg->name.str= NULL;
		    cfg->action.text= NULL;
		}
	    }
	    else {
		uWarning("Assignment to unknown variable \"%s\"\n",cfg->name);
		uAction("Ignored\n");
	    }
	}
	else if (cfg->entry_type==EventDef) switch (cfg->event_type) {
	    case XkbBellNotify:
		if (name!=NULL)	cfg->name.atom= XInternAtom(dpy,name,False);
		else 		cfg->name.atom= None;
		if (name) uFree(name);
		break;
	    case XkbAccessXNotify:
		priv= 0;
		if (name==NULL)		
		     priv= XkbAllNewKeyboardEventsMask;
		else if (uStrCaseEqual(name,"skpress"))	
		     priv= XkbAXN_SKPressMask;
		else if (uStrCaseEqual(name,"skaccept"))
		     priv= XkbAXN_SKAcceptMask;
		else if (uStrCaseEqual(name,"skreject"))
		     priv= XkbAXN_SKRejectMask;
		else if (uStrCaseEqual(name,"skrelease"))
		     priv= XkbAXN_SKReleaseMask;
		else if (uStrCaseEqual(name,"bkaccept"))
		     priv= XkbAXN_BKAcceptMask;
		else if (uStrCaseEqual(name,"bkreject"))
		     priv= XkbAXN_BKRejectMask;
		else if (uStrCaseEqual(name,"warning"))
		     priv= XkbAXN_AXKWarningMask;
		if (name)	uFree(name);
		cfg->name.priv= priv;
		break;
	    case XkbActionMessage:
		/* nothing to do */
		break;
	}
	eventMask|= (1L<<cfg->event_type);
	cfg= cfg->next;
    }
    while ((config)&&(config->entry_type!=EventDef)) {
	CfgEntryPtr next;
	if (config->name.str)		uFree(config->name.str);
	if (config->action.text)	uFree(config->action.text);
	config->name.str= 	NULL;
	config->action.text=	NULL;
	next= 			config->next;
	uFree(config);
	config= next;
    }
    cfg= config;
    while ((cfg!=NULL)&&(cfg->next!=NULL)) {
	CfgEntryPtr next;
	next= cfg->next;
	if (next->entry_type!=EventDef) {
	    if (next->name.str)		uFree(config->name.str);
	    if (next->action.text)	uFree(config->action.text);
	    next->name.str=		NULL;
	    next->action.text=		NULL;
	    cfg->next= 			next->next;
	    next->next=			NULL;
	    uFree(next);
	}
	else cfg= next;
    }
    return;
}

static CfgEntryPtr
FindMatchingConfig(XkbEvent *ev)
{
CfgEntryPtr	cfg,dflt;

    dflt= NULL;
    for (cfg= config;(cfg!=NULL);cfg=cfg->next) {
	if ((ev->type!=xkbEventCode)||(cfg->event_type!=ev->any.xkb_type))
	    continue;
	switch (ev->any.xkb_type) {
	   case XkbBellNotify:
		if (ev->bell.name==cfg->name.atom)
		    return cfg;
		else if ((cfg->name.atom==None)&&(dflt==NULL))
		    dflt= cfg;
		break;
	    case XkbAccessXNotify:
		if (cfg->name.priv&(1L<<ev->accessx.detail))
		    return cfg;
		break;
	    case XkbActionMessage:
		if (cfg->name.str==NULL)
		    dflt= cfg;
		else if (strncmp(cfg->name.str,ev->message.message,
						XkbActionMessageLength)==0)
		    return cfg;
		break;
	   default:
		uInternalError("Can't handle type %d XKB events yet, Sorry.\n");
		break;
	}
    }
    return dflt;
}

static Bool
ProcessMatchingConfig(XkbEvent *ev)
{
CfgEntryPtr	cfg;
char		buf[1024],*cmd;
int		ok;

    cfg= FindMatchingConfig(ev);
    if (!cfg) {
	if (verbose)
	    PrintXkbEvent(stdout,ev);
	return False;
    }
    if (cfg->action.type==UnknownAction) {
	if (cfg->action.text==NULL)
	    cfg->action.type= NoAction;
	else if (cfg->action.text[0]=='!') {
	    char *tmp;
	    cfg->action.type= ShellAction;
	    tmp= uStringDup(&cfg->action.text[1]);
	    uFree(cfg->action.text);
	    cfg->action.text= tmp;
	}
	else cfg->action.type= SoundAction;
    }
    switch (cfg->action.type) {
	case NoAction:
	    return True;
	case EchoAction:
	    if (cfg->action.text!=NULL) {
		sprintf(buf,"%s",cfg->action.text);
		cmd= SubstituteEventArgs(buf,ev);
		printf("%s",cmd);
	    }
	    return True;
	case PrintEvAction:
	    PrintXkbEvent(stdout,ev);
	    return True;
	case ShellAction:
	    if (cfg->action.text==NULL) {
		uWarning("Empty shell command!\n");
		uAction("Ignored\n");
		return True;
	    }
	    cmd= cfg->action.text;
	    break;
	case SoundAction:
	    if (cfg->action.text==NULL) {
		uWarning("Empty sound command!\n");
		uAction("Ignored\n");
		return True;
	    }
	    sprintf(buf,"%s %s%s",soundCmd,soundDir,cfg->action.text);
	    cmd= buf;
	    break;
	default:
	    uInternalError("Unknown error action type %d\n",cfg->action.type);
	    return False;
    }
    cmd= SubstituteEventArgs(cmd,ev);
    if (verbose)
	uInformation("Executing shell command \"%s\"\n",cmd);
    ok= (system(cmd)==0);
    return ok;
}

/***====================================================================***/

int
main(int argc, char *argv[])
{
FILE 	*	file;
static char 	buf[1024];
XkbEvent	ev;
Bool		ok;


    yyin = stdin;
    uSetEntryFile(NullString);
    uSetDebugFile(NullString);
    uSetErrorFile(NullString);

    if (!parseArgs(argc,argv))
	exit(1);
    file= NULL;
    XkbInitAtoms(NULL);
    if (cfgFileName==NULL) {
	char *home;
	home= (char *)getenv("HOME");
	sprintf(buf,DFLT_XKBEVD_CONFIG,(home?home:""));
	cfgFileName= buf;
    }
    if (uStringEqual(cfgFileName,"-")) {
	static char *in= "stdin";
	file= stdin;
	cfgFileName= in;
    }
    else {
	file= fopen(cfgFileName,"r");
	if (file==NULL) { /* no personal config, try for a system one */
	    if (cfgFileName!=buf) { /* user specified a file.  bail */
		uError("Can't open config file \"%s\n",cfgFileName);
		uAction("Exiting\n");
		exit(1);
	    }
	    sprintf(buf,DFLT_SYS_XKBEVD_CONFIG,DFLT_XKB_CONFIG_ROOT);
	    file= fopen(cfgFileName,"r");
	    if (file==NULL && !eventMask) {
		if (verbose) {
		    uError("Couldn't find a config file anywhere\n");
		    uAction("Exiting\n");
		    exit(1);
		}
		exit(0);
	    }
	}
    }

    if (background) {
	if (fork()!=0) {
	    if (verbose) 
		uInformation("Running in the background\n");
	    exit(0);
	}
    }
    dpy= GetDisplay(argv[0],dpyName,&xkbOpcode,&xkbEventCode);
    if (!dpy)
	goto BAILOUT;
    ok= True;
    setScanState(cfgFileName,1);
    CFGParseFile(file);
    if (!config && !eventMask) {
	uError("No configuration specified in \"%s\"\n",cfgFileName);
	goto BAILOUT;
    }
    if (eventMask==0) {
	uError("No events to watch in \"%s\"\n",cfgFileName);
	goto BAILOUT;
    }
    if (!XkbSelectEvents(dpy,XkbUseCoreKbd,eventMask,eventMask)) {
	uError("Couldn't select desired XKB events\n");
	goto BAILOUT;
    }
    xkb= XkbGetKeyboard(dpy,XkbGBN_AllComponentsMask,XkbUseCoreKbd);
    if (eventMask&XkbBellNotifyMask) {
	unsigned ctrls,vals;
	if (verbose)
	   uInformation("Temporarily disabling the audible bell\n");
	if (!XkbChangeEnabledControls(dpy,XkbUseCoreKbd,XkbAudibleBellMask,0)) {
	    uError("Couldn't disable audible bell\n");
	    goto BAILOUT;
	}
	ctrls= vals= XkbAudibleBellMask;
	if (!XkbSetAutoResetControls(dpy,XkbAudibleBellMask,&ctrls,&vals)) {
	    uWarning("Couldn't configure audible bell to reset on exit\n");
	    uAction("Audible bell might remain off\n");
	}
    }
    if (soundCmd==NULL)	soundCmd= DFLT_SOUND_CMD;
    if (soundDir==NULL)	soundDir= DFLT_SOUND_DIR;
    XkbStdBellEvent(dpy,None,0,XkbBI_ImAlive);
    while (1) {
	XNextEvent(dpy,&ev.core);
	if ((!ProcessMatchingConfig(&ev))&&(ev.type==xkbEventCode)&&
					(ev.any.xkb_type==XkbBellNotify)) {
	    XkbForceDeviceBell(dpy,ev.bell.device,
				ev.bell.bell_class,ev.bell.bell_id,
				ev.bell.percent);
	}
    }

    XCloseDisplay(dpy);
    return (ok==0);
BAILOUT:
    uAction("Exiting\n");
    if (dpy!=NULL)
	XCloseDisplay(dpy);
    exit(1);
}
